src/common/cockpitconf.c

```diff
@@ -21,92 +21,152 @@
+{
+  if (nmemb >= SIZE_MAX / size)
+    err (1, "reallocarr: overflow (nmemb %zu)", nmemb);
+  return realloc (ptr, nmemb * size);
+}
 
-  if (!g_key_file_load_from_file (key_file, file_path, G_KEY_FILE_NONE, &error))
+/* See https://developer.gnome.org/glib/stable/glib-Key-value-file-parser.html for the spec,
+ * except that cockpit.conf is case insensitive for section and key names; so store them in lower case */
+static bool
+load_key_file (const char *file_path)
+{
+  FILE *f = NULL;
+  char *cur_section = NULL;
+  regex_t re_section, re_keyval, re_ignore;
+  char *line = NULL;
+  bool ret = true;
+  size_t line_size = 0;
+  ssize_t line_len;
+
+  cockpit_conf_loaded = true;
+
+  f = fopen (file_path, "r");
+  if (!f)
     {
-      if (g_error_matches (error, G_FILE_ERROR, G_FILE_ERROR_NOENT))
-        goto out;
-
-      g_message ("couldn't load configuration file: %s: %s", file_path, error->message);
-      ret = FALSE;
-      goto out;
+      if (errno != ENOENT)
+        warnx ("couldn't load configuration file: %s: %m\n", file_path);
+      return false;
     }
 
-  groups = g_key_file_get_groups (key_file, NULL);
-  for (i = 0; groups[i] != NULL; i++)
+  assert (regcomp (&re_section, "^[[:space:]]*\\[ *([^][:space:]]+) *\\][[:space:]]*$", REG_EXTENDED|REG_NEWLINE) == 0);
+  assert (regcomp (&re_keyval, "^[[:space:]]*([^=[:space:]]+)[[:space:]]*=[[:space:]]*(.*)$", REG_EXTENDED|REG_NEWLINE) == 0);
+  assert (regcomp (&re_ignore, "^[[:space:]]*(#.*)?$", REG_EXTENDED|REG_NOSUB) == 0);
+
+  while ((line_len = getline (&line, &line_size, f)) >= 0)
     {
-      gchar **keys;
-      section = g_hash_table_lookup (cockpit_conf, groups[i]);
-      if (section == NULL)
+      regmatch_t matches[3];
+
+      if (regexec (&re_section, line, 3, matches, 0) == 0)
         {
-          section = g_hash_table_new_full (cockpit_str_case_hash, cockpit_str_case_equal, g_free, g_free);
-          g_hash_table_insert (cockpit_conf, g_strdup (groups[i]), section);
+          free (cur_section);
+          cur_section = strlower (strndup (line + matches[1].rm_so, matches[1].rm_eo - matches[1].rm_so));
         }
 
-      keys = g_key_file_get_keys (key_file, groups[i], NULL, &error);
-      g_return_val_if_fail (error == NULL, FALSE);
+      else if (regexec (&re_keyval, line, 3, matches, 0) == 0)
+        {
+          Entry *e;
+
+          if (!cur_section)
+            {
+              warnx ("%s: key=val line not in any section: %s", file_path, line);
+              ret = false;
+              break;
+            }
+
+          e = malloc (sizeof (Entry));
+          e->section = strdup (cur_section);
+          e->key = strlower (strndup (line + matches[1].rm_so, matches[1].rm_eo - matches[1].rm_so));
+          e->value = strndup (line + matches[2].rm_so, matches[2].rm_eo - matches[2].rm_so);
+          e->strv_cache = NULL;
+          /* prepend new Entry to cockpit_conf; that way, later values win over earlier ones in a forward search */
```

Comment:
This would break compatibility. The current code accepts duplicate keys, and the last one wins (or  maybe that's not strictly defined). You can play around with src/ws/mock-config/cockpit/cockpit.conf and ./test-config to check.